### PR TITLE
fix: handle empty shards

### PR DESCRIPTION
**Pull request by cristianmtr**
Fixes https://github.com/jina-ai/jina/issues/1229

### CODE REVIEW

File: jina/executors/decorators.py [hunk 1]
```diff
@@ -247,6 +247,10 @@ def arg_wrapper(*args, **kwargs):
             if b_size is None:
                 return func(*args, **kwargs)
 
+            # catch empty shards
+            if data is None:
```
Comment (JoanFM) [inline jina/executors/decorators.py:251]: I think is not the place to add the fix, I am not sure the return expected is an `array`. I think it needs to be added at `indexer`

Comment (cristianmtr) [reply jina/executors/decorators.py:251]: Well the crash happens at the batch_iterator level

```
    for b in batch_iterator(data[:total_size], b_size, split_over_axis, yield_slice=yield_slice):
TypeError: 'NoneType' object is not subscriptable
```

And I assume it will be several indexers that will call this.

And when looking at the indexer query https://github.com/jina-ai/jina/blob/efdb11342536426b78d1c5d448c64a8f3634027a/jina/executors/indexers/vector.py#L263

```
        if self.metric not in {'cosine', 'euclidean'} or self.backend == 'scipy':
            dist = self._cdist(keys, self.query_handler)
        elif self.metric == 'euclidean':
            _keys = _ext_A(keys)
            dist = self._euclidean(_keys, self.query_handler)
        elif self.metric == 'cosine':
            _keys = _ext_A(_norm(keys))
            dist = self._cosine(_keys, self.query_handler)
        else:
            raise NotImplementedError(f'{self.metric} is not implemented')
```

I see that I would need to add it for each `if` substatement, right?

Also: still a draft. Need to add the tests.

As for return type: if you look at the function that would return after the `for` we can look at this function https://github.com/jina-ai/jina/blob/f0809e32db805401b2768aef7e8eec9c4ebae83d/jina/executors/decorators.py#L179, which seems to return an array

Comment (JoanFM) [reply jina/executors/decorators.py:251]: yes, maybe some change also needs to be there for `batch_iterator` but it will also affect the `indexer` executor.

Comment (JoanFM) [reply jina/executors/decorators.py:251]: but it is not necessaritly an `np.array`, it may also a simple `array` I think

File: tests/unit/executors/test_decorators.py [hunk 2]
```diff
@@ -9,6 +9,8 @@
 cur_dir = os.path.dirname(os.path.abspath(__file__))
 
 
+# TODO add test empty index with search
```
Comment (JoanFM) [inline tests/unit/executors/test_decorators.py:12]: this TODO should not be here right?

Comment (JoanFM) [reply tests/unit/executors/test_decorators.py:12]: the tests would belong somewhere else I think?

File: jina/executors/decorators.py (hunk 1 above)
Comment (JoanFM) [reply jina/executors/decorators.py:251]: I think I was wrong it may be actually a bug at batching level. I am howevrr not sure we can assume it returns an np.ndarray. batching decorator has too many things, if you find a pattern to abstract in another function it can also be good

Comment (cristianmtr) [reply jina/executors/decorators.py:251]: I think you're looking at an older version. After discussing it with @maximilianwerk we said the batching should just raise the error. And it's up to the Indexers / modules to know what sort of data and format to default to

File: tests/unit/executors/test_decorators.py (hunk 2 above)
Comment (cristianmtr) [reply tests/unit/executors/test_decorators.py]: A very basic version is here https://github.com/jina-ai/jina/pull/1340/files#diff-f72cb602d29fda8028f46b7b7d9940710c62d5b7fb95966f1fa58f6c924cc2d2R242

File: jina/executors/decorators.py (hunk 1 above)
Comment (JoanFM) [reply jina/executors/decorators.py:251]: I think it is not the way to go, because this is not the data passed but the data returned. I think what batching is responsible is to handle data that comes and goes, if it does not receive any data it should just return None.

raising an error is a big statement in my opinion, it should be a natural thing for batching to handle empty data.

Comment (maximilianwerk) [reply jina/executors/decorators.py:251]: I disagree. Batching has no clue about the type of the returned value and cannot guess it. If the type is `Option[Any]`, `None` is fine. Anyhow, if the type is not `Option`, `None` does not make any sense, so either we call the underlying function with an empty `np.ndarray` and let it be handled there or we raise an `Exception`.

File: jina/executors/indexers/vector.py [hunk 3]
```diff
@@ -306,3 +310,9 @@ def _cdist(self, *args, **kwargs):
             return cdist(*args, **kwargs, metric=self.metric)
         except ModuleNotFoundError:
             raise ModuleNotFoundError(f'your metric {self.metric} requires scipy, but scipy is not found')
+
+    def _is_empty(self):
```
Comment (JoanFM) [inline jina/executors/indexers/vector.py:314]: I think the emptiness is extracted from the `_size` and not from the `query_handler` being None

File: tests/unit/executors/indexers/test_numpyindexer.py [hunk 4]
```diff
@@ -160,3 +160,29 @@ def test__get_sorted_top_k(batch_size, num_docs, top_k):
...
+
+@pytest.mark.parametrize('batch_size, compress_level', [(None, 0), (None, 1), (2, 0), (2, 1)])
+def test_numpy_indexer_empty_data(batch_size, compress_level, test_metas):
+    np.random.seed(500)
+    num_data = 100
+    num_dim = 64
+    num_query = 10
+    vec_idx = np.random.randint(0, high=num_data, size=[num_data])
+    # empty data
+    vec = np.empty((0, 0))
```
Comment (JoanFM) [inline tests/unit/executors/indexers/test_numpyindexer.py:173]: Are we sure this reflects the problem we faced? Empty shards are empty because they indexed nothing not because they received an empty index query right?

**Review by JoanFM — CHANGES_REQUESTED** [review]

**Comment by JoanFM** [comment]
I think this ticket requires to add an integration test when a `Flow` with an `Index Pod` with `sharding=2` indexes only one document, and then trying to retrieve that document still works.

File: jina/executors/indexers/vector.py (hunk 3 above)
Comment (cristianmtr) [reply jina/executors/indexers/vector.py:314]: Got it. The problem was that we were passing keys even though there wasn't any data. Fixed that now

File: tests/unit/executors/indexers/test_numpyindexer.py (hunk 4 above)
Comment (cristianmtr) [reply tests/unit/executors/indexers/test_numpyindexer.py]: Got it

Comment (JoanFM) [reply tests/unit/executors/indexers/test_numpyindexer.py]: it may still be an interesting usecase to test though

File: jina/executors/indexers/vector.py [hunk 5]
```diff
@@ -306,3 +310,10 @@ def _cdist(self, *args, **kwargs):
...
```
Comment (JoanFM) [inline jina/executors/indexers/vector.py:314]: do we still need `_is_empty`?

**Review by JoanFM — CHANGES_REQUESTED** [review]
Missing integration test showcasing issue is fixed

File: tests/integration/issues/github_1229/test_sharding_empty_index.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+def test_sharding_empty_index(random_workspace_name):
+    print("WORKSPACE =", random_workspace_name)
+    os.environ['JINA_TEST_1229_WORKSPACE'] = os.path.abspath(random_workspace_name)
+    num_shards = 2
+
+    f = Flow() \
+        .add(
+        uses='vectorindexer.yml',
+        shards=1,
```
Comment (JoanFM) [inline tests/integration/issues/github_1229/test_sharding_empty_index.py:19]: u need to index with 2 shards. I guess it is still on progress right?

File: tests/unit/executors/test_decorators.py [hunk 7]
```diff
@@ -5,6 +5,7 @@
 
 from jina.executors.decorators import as_update_method, as_train_method, as_ndarray, batching, \
     require_train, store_init_kwargs, batching_multi_input
+from jina.executors.indexers.vector import _ext_B
```
Comment (JoanFM) [inline tests/unit/executors/test_decorators.py:8]: I guess this is not needed anymore?

File: jina/executors/indexers/vector.py (hunk 5 above)
Comment (cristianmtr) [reply jina/executors/indexers/vector.py]: Removed

File: tests/integration/issues/github_1229/test_sharding_empty_index.py (hunk 6 above)
Comment (maximilianwerk) [reply tests/integration/issues/github_1229/test_sharding_empty_index.py]: Yes it was to shove the code to me for debugging.

Comment (cristianmtr) [reply tests/integration/issues/github_1229/test_sharding_empty_index.py]: Yep, done now.

File: tests/unit/executors/test_decorators.py (hunk 7 above)
Comment (cristianmtr) [reply tests/unit/executors/test_decorators.py]: Removed.

**Comment by cristianmtr** [comment]
> I think this ticket requires to add an integration test when a `Flow` with an `Index Pod` with `sharding=2` indexes only one document, and then trying to retrieve that document still works.

Yep, added now

File: tests/integration/issues/github_1229/test_sharding_empty_index.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    data = []
+    for i in range(num_docs):
+        doc = Document()
+        doc.content = f'random {i}'
+        doc.embedding = np.array([i])
+        doc.update_id()
+        data.append(doc)
+
+    with f:
+        f.index(data)
```
Comment (JoanFM) [inline tests/integration/issues/github_1229/test_sharding_empty_index.py:32]: to make more explicit that only one shard will have data, add `f.index(data, batch_size=2)` like this u are sure that the request is not split in 2

Comment (JoanFM) [reply tests/integration/issues/github_1229/test_sharding_empty_index.py:32]: Just to make sure, have you locally tested this fails without ur changes in current master?

Comment (cristianmtr) [reply tests/integration/issues/github_1229/test_sharding_empty_index.py:32]: Hm good point. It seems like the integration test is passing even without my change. However, the multires lyrics example at this commit (https://github.com/jina-ai/examples/blob/96863bec5de24b589fab943bad92fde5328a3ef0/multires-lyrics-search/tests/test_multireslyrics.py#L18) does fail without it and pass with it. That's the original experiment where we noticed the issue

Comment (JoanFM) [reply tests/integration/issues/github_1229/test_sharding_empty_index.py:32]: Try with `batch_size=2`

Comment (cristianmtr) [reply tests/integration/issues/github_1229/test_sharding_empty_index.py:32]: I reduced the nr of docs to 1 and it still passed

Comment (JoanFM) [reply tests/integration/issues/github_1229/test_sharding_empty_index.py:32]: then we need to understand why this is passing in master

**Comment by cristianmtr** [comment]
Just a FYI I am getting empty content on the matches for the docs:

```python
        for d in result.docs:
            print(f'document: {d.content}')
            print(f'matches = {[m.content for m in d.matches]}')
```

```
document: query 0
matches = [None]
document: query 1
matches = [None]
document: query 2
matches = [None]
document: query 3
matches = [None]
document: query 4
matches = [None]
document: query 5
matches = [None]
document: query 6
matches = [None]
document: query 7
matches = [None]
document: query 8
matches = [None]
document: query 9
matches = [None]
```

File: tests/integration/issues/github_1229/test_sharding_empty_index.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
+import os
+
+import numpy as np
+
+from jina import Flow, Document
+
+PORT = 45888
```
Comment (bwanglzu) [inline tests/integration/issues/github_1229/test_sharding_empty_index.py:7]:
```python
from jina.helper import random_port
```

File: tests/integration/issues/github_1229/test_sharding_empty_index.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+        f.index(data)
+
+    num_query = 10
+    query = []
+    for i in range(num_query):
+        doc = Document()
+        doc.content = f'query {i}'
+        doc.embedding = np.array([i])
+        doc.update_id()
+        query.append(doc)
```
Comment (bwanglzu) [inline tests/integration/issues/github_1229/test_sharding_empty_index.py:42]: maybe you can use `with Document() as doc` and you do not necessarily to call `doc.update_id()`

**Review by bwanglzu — CHANGES_REQUESTED** [review]
not sure, do you see it's related to #1372 ?

**Comment by cristianmtr** [comment]
The integration test in `test_sharding_empty_index.py` does seem to be passing even without the changes introduced in the Indexer. But it's passing with the empty results from above. 

However, the unit test `test_numpy_indexer_empty_data` does fail without the changes introduced in this PR.

File: tests/integration/issues/github_1229/test_sharding_empty_index.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+    def callback(result):
+        for d in result.docs:
+            print(f'document: {d.content}')
+            print(f'matches = {[m.content for m in d.matches]}')
+        assert result is not None
```
Comment (maximilianwerk) [inline tests/integration/issues/github_1229/test_sharding_empty_index.py:48]: That `assert` is never reached on master. Better might be something along:
```
    callback_was_called = False

    def callback(result):
        callback_was_called = True
        for d in result.docs:
            print(f'document: {d.content}')
            print(f'matches = {[m.content for m in d.matches]}')
        assert result is not None

    with f:
        f.search(query, output_fn=callback)
    assert callback_was_called
```
I am not sure, whether there is a more pytesty way to do this. But that's what needs testing.

File: tests/integration/issues/github_1229/test_sharding_empty_index.py (hunk 8 above)
Comment (maximilianwerk) [reply tests/integration/issues/github_1229/test_sharding_empty_index.py:33]: See my comment below

File: tests/integration/issues/github_1229/test_sharding_empty_index.py [hunk 12]
```diff
@@ -0,0 +1,51 @@
+import os
+
+import numpy as np
+
+from jina import Flow, Document
+
+PORT = 45888
+
+
+def test_sharding_empty_index(random_workspace_name):
```
Comment (maximilianwerk) [inline tests/integration/issues/github_1229/test_sharding_empty_index.py:10]: I showed you the `random_workspace_name`, but apparently, either we use plain pytest `tmpdir` fixture or we use `test_metas`. Since the second one is a bit overkill, I'd say, let' go for `tmpdir`. Main reason: The created directories are in the operating system temporary directory and not in the jina main directory.

Comment (cristianmtr) [reply tests/integration/issues/github_1229/test_sharding_empty_index.py]: Done

File: tests/integration/issues/github_1229/test_sharding_empty_index.py (hunk 11 above)
Comment (cristianmtr) [reply tests/integration/issues/github_1229/test_sharding_empty_index.py]: It is reached, because it prints the documents and it is triggered in debugger.

The problem in the case above is that `callback_was_called` is not referenced as `global` within the callback. See my new commit.

```python
callback_was_called = False


def test_sharding_empty_index(tmpdir):
...

    def callback(result):
        global callback_was_called
        callback_was_called = True
        for d in result.docs:
            print(f'document: {d.content}')
            print(f'matches = {[m.content for m in d.matches]}')

    with f:
        f.search(query, output_fn=callback)
    assert callback_was_called

```

This passes, but I still get `None` in the matches content.

Comment (JoanFM) [reply tests/integration/issues/github_1229/test_sharding_empty_index.py]: Well u were printing the `document` and `matches` right? so it was called

Comment (cristianmtr) [reply tests/integration/issues/github_1229/test_sharding_empty_index.py]: Yes

File: tests/integration/issues/github_1229/test_sharding_empty_index.py [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+    num_query = 10
+    query = []
+    for i in range(num_query):
+        with Document() as doc:
+            doc.content = f'query {i}'
+            doc.embedding = np.array([i])
+            query.append(doc)
+
+    def callback(result):
+        global callback_was_called
```
Comment (JoanFM) [inline tests/integration/issues/github_1229/test_sharding_empty_index.py:53]: This is something that I find interesting:
1. On one side I hate to see the `global` variable, but I like this idea that guarantees that we do not have false positivies

However, 

we could pass to the callback some `container` to be switched. (We could wrap a boolean in a class that it can be switched in the callback). This could even be made a `fixture` available for every test (which may not be available for the global variable)

**Comment by cristianmtr** [comment]
Adding `polling='all'` fixes it. But then, when I remove the changes from decorator and vector, I cannot reproduce the error from multires lyrics

File: jina/executors/indexers/vector.py [hunk 14]
```diff
@@ -272,6 +272,10 @@ def query(self, keys: 'np.ndarray', top_k: int, *args, **kwargs) -> Tuple['np.nd
             Distance (the smaller the better) is returned, not the score.
 
         """
+        if self.size == 0:
+            query_count = len(keys)
+            result_count = 0
+            return np.empty((query_count, result_count)), np.empty((query_count, result_count))
```
Comment (hanxiao) [inline jina/executors/indexers/vector.py:278]: this is pretty dangerous, `numpy.empty` is like `memalloc` it assigns a memory but not initializing it, resulting in a *random* array. How can one tell if this is legit or not later?

Comment (hanxiao) [reply jina/executors/indexers/vector.py:278]:
```
>>> import numpy as np
>>> np.empty([3, 3])
array([[ 2.00000000e+000,  2.00000000e+000, -3.95252517e-323],
       [ 0.00000000e+000,  2.12199579e-314,  0.00000000e+000],
       [ 0.00000000e+000,  0.00000000e+000,  1.77229088e-310]])
>>>
```

**Review by hanxiao — CHANGES_REQUESTED** [review]
Not sure why `numpy.empty` is used here

File: jina/executors/indexers/vector.py (hunk 14 above)
Comment (cristianmtr) [reply jina/executors/indexers/vector.py:278]: Well in this case, since one of the dims is `0` it won't be anything. But I understand your general point. I guess it should be `zeros`?

File: tests/unit/executors/indexers/test_numpyindexer.py [hunk 15]
```diff
@@ -160,3 +160,25 @@ def test__get_sorted_top_k(batch_size, num_docs, top_k):
...
+        indexer.touch()
+        indexer.save()
+        assert os.path.exists(indexer.index_abspath)
+        save_abspath = indexer.save_abspath
+
+    with BaseIndexer.load(save_abspath) as indexer:
+        assert isinstance(indexer, NumpyIndexer)
+        idx, dist = indexer.query(query, top_k=4)
+        assert idx.shape == dist.shape
+        assert idx.shape == (num_query, 0)
```
Comment (hanxiao) [inline tests/unit/executors/indexers/test_numpyindexer.py:184]: so is the previous `numpy.empty` designed for complying with this?

Comment (cristianmtr) [reply tests/unit/executors/indexers/test_numpyindexer.py:184]: That's just the `.shape`, but yes.

Comment (hanxiao) [reply tests/unit/executors/indexers/test_numpyindexer.py:184]: in #1229 you mentioned
> If a shard is empty, is should still return an empty list of results.

so the desired behavior is to return an empty ndarray with the shape `[num_query, 0]`?

Comment (hanxiao) [reply tests/unit/executors/indexers/test_numpyindexer.py:184]: is there some reason to keep the `axis=0`? or what the choice between
```python
>>> np.empty([0, 0])
array([], shape=(0, 0), dtype=float64)
>>> np.empty([3, 0])
array([], shape=(3, 0), dtype=float64)
>>> np.empty([0])
array([], dtype=float64)
```

Comment (cristianmtr) [reply tests/unit/executors/indexers/test_numpyindexer.py:184]: > in #1229 you mentioned
> 
> > If a shard is empty, is should still return an empty list of results.
> 
> so the desired behavior is to return an empty ndarray with the shape `[num_query, 0]`?

Yes

Comment (cristianmtr) [reply tests/unit/executors/indexers/test_numpyindexer.py:184]: I think we want this, just so that the shapes still asign (nr of queries - first dim)

```python
>>> np.empty([3, 0])
array([], shape=(3, 0), dtype=float64)
```

Comment (hanxiao) [reply tests/unit/executors/indexers/test_numpyindexer.py:184]: can we `return None, None`?

Comment (hanxiao) [reply tests/unit/executors/indexers/test_numpyindexer.py:184]: what I don't like about this part is: as a user of this API, you don't know if the indexer is empty *until you inspect at axis=1*, and then say "oh, axis=1 is 0, so it must be empty!" - which isn't an obvious connection unless you documented in that way.

Return `None, None` would be much more direct.

File: tests/integration/issues/github_1229/test_sharding_empty_index.py [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+    for i in range(num_docs):
+        with Document() as doc:
+            doc.content = f'data {i}'
+            doc.embedding = np.array([i])
+            data.append(doc)
+
+    with f:
+        f.index(data)
+
+    f.close()
```
Comment (nan-wang) [inline tests/integration/issues/github_1229/test_sharding_empty_index.py:55]: Not necessary. `with` will call the context manager to close it
```suggestion
```

File: tests/unit/executors/indexers/test_numpyindexer.py (hunk 15 above)
Comment (cristianmtr) [reply tests/unit/executors/indexers/test_numpyindexer.py:184]: But, for example, in the case of sharding, where we want to merge results from two different shards, if one is empty, it's OK, they can merge. If one is `None, None`, error. Right?

File: tests/integration/issues/github_1229/test_sharding_empty_index.py [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+            doc.content = f'query {i}'
+            doc.embedding = np.array([i])
+            query.append(doc)
+
+    def callback(result):
+        global callback_was_called
+        callback_was_called = True
+        assert len(result.docs) == num_query
+        for d in result.docs:
+            print(f'document: {d.content} with {len(d.matches)} matches')
```
Comment (JoanFM) [inline tests/integration/issues/github_1229/test_sharding_empty_index.py:70]: may we remove all these prints?

**Review by JoanFM — APPROVED** [review]

**Review by JoanFM — CHANGES_REQUESTED** [review]
The empties should be handled differently?

File: tests/integration/issues/github_1229/test_sharding_empty_index.py (hunk 17 above)
Comment (cristianmtr) [reply tests/integration/issues/github_1229/test_sharding_empty_index.py]: Done

File: tests/unit/executors/indexers/test_numpyindexer.py (hunk 15 above)
Comment (hanxiao) [reply tests/unit/executors/indexers/test_numpyindexer.py:184]: then this logic should be put into driver. which merge are we talking about? can you point out the function?

**Review by hanxiao — CHANGES_REQUESTED** [review]
the `None, None` handling should be put into driver, which merge function are we talking about, can someone points out the function?

**Comment by cristianmtr** [comment]
> the `None, None` handling should be put into driver, which merge function are we talking about, can someone points out the function?

How about returning something like

```python
        if self.size == 0:
            query_count = len(keys)
            result = np.array([[] for _ in range(query_count)])
            return result, result
```

So we maintain compatibility with the interface?

Not sure what would break if we switch to `None, None`, running all the tests now to see

**Comment by maximilianwerk** [comment]
> > the `None, None` handling should be put into driver, which merge function are we talking about, can someone points out the function?
> 
> How about returning something like
> 
> ```python
>         if self.size == 0:
>             query_count = len(keys)
>             result = np.array([[] for _ in range(query_count)])
>             return result, result
> ```
> 
> So we maintain compatibility with the interface?
> 
> Not sure what would break if we switch to `None, None`, running all the tests now to see

The way `_merge_all` is constructed, returning `None, None` should be fine. Haven't thought about it that way beforehand. I rather thought we want to stick with the function interface. But I agree, that if it just works, it makes sense on this level.

**Comment by cristianmtr** [comment]
> > > the `None, None` handling should be put into driver, which merge function are we talking about, can someone points out the function?
> > 
> > 
> > How about returning something like
> > ```python
> >         if self.size == 0:
> >             query_count = len(keys)
> >             result = np.array([[] for _ in range(query_count)])
> >             return result, result
> > ```
> > 
> > 
> > So we maintain compatibility with the interface?
> > Not sure what would break if we switch to `None, None`, running all the tests now to see
> 
> The way `_merge_all` is constructed, returning `None, None` should be fine. Haven't thought about it that way beforehand. I rather thought we want to stick with the function interface. But I agree, that if it just works, it makes sense on this level.

Done, returning `None, None` now. Adapted interface to be `Optional[...]`

**Review by hanxiao — APPROVED** [review]
👍

**Comment by hanxiao** [comment]
the most discussed PR so far?

File: tests/unit/executors/indexers/test_numpyindexer.py [hunk 18]
```diff
@@ -160,3 +160,25 @@ def test__get_sorted_top_k(batch_size, num_docs, top_k):
...
+        indexer.touch()
+        indexer.save()
+        assert os.path.exists(indexer.index_abspath)
+        save_abspath = indexer.save_abspath
+
+    with BaseIndexer.load(save_abspath) as indexer:
+        assert isinstance(indexer, NumpyIndexer)
+        idx, dist = indexer.query(query, top_k=4)
+        assert idx == None
+        assert idx == None
```
Comment (nan-wang) [inline tests/unit/executors/indexers/test_numpyindexer.py:184]: Fix the duplicates.
```
assert dist == None
```

Comment (cristianmtr) [reply tests/unit/executors/indexers/test_numpyindexer.py:184]: https://github.com/jina-ai/jina/pull/1396
